### PR TITLE
Image view

### PR DESCRIPTION
**Pull request by Juliotati**
Updated the code in the original closed PR #24 

Reason for removing the `canLaunchUrl` check

Screenshot 2023-05-29 at 14 41 47

### CODE REVIEW

**Comment by Juliotati** [comment]
wrong commit message there, meant to say:
- image view from #24

**Comment by mkobuolys** [comment]
> @mkobuolys wrong commit message there, meant to say:
> 
> * image view from [Add better image view #24](https://github.com/mkobuolys/flutter-design-patterns/pull/24)
> 
> Although it bring a good point, that the app could be put in the process for publishing now? I guess the only thing left is:
> 
> * App icon
> * and double check on the `packageId` and display name

Honestly, I don't think that this is a good point. To me, the app is not worth publishing as it is now. At the moment, Flutter Design Patterns is just a lightweight library of design pattern examples implemented in Dart and... that's it?

Here is my take on this project:
1) There is a dedicated domain https://www.flutterdesignpatterns.com where this app is deployed on each code change in the main branch. I would rather keep it this way than semi-automate the release process to the stores.
2) To expand on the point mentioned in the issue about offline learning: I would rather spend some time preparing the app as a fully offline-first PWA application than build an Android/iOS app.

Having these points in mind, I think about removing ios/android directories from the project and focusing on the web version. Also, reducing the maintenance overhead (basically, what you did with this PR by updating the Android project).

**Comment by Juliotati** [comment]
@mkobuolys I agree seems reasonable, I mainly use the web for everything, and i would use the PWA if i ever need to check on something when can't access the app through a desktop.

**Comment by mkobuolys** [comment]
Anyway, platform-specific changes are out of scope for this issue/PR, I would like to take a look at them myself or even remove the platforms as mentioned above. Please, revert the changes and only leave Dart code changes to proceed with the review.

**Review by Juliotati — COMMENTED** [review]
@mkobuolys  done ✨

File: lib/modules/design_pattern_details/widgets/rich_markdown_body.dart [hunk 1]
```diff
@@ -93,6 +94,19 @@ class RichMarkdownBody extends StatelessWidget {
       ),
       onTapLink: (_, link, __) => UrlLauncher.launchUrl(link ?? ''),
       syntaxHighlighter: DartSyntaxHighlighter(context),
+      imageBuilder: (uri, _, __) {
```
Comment (mkobuolys) [inline lib/modules/design_pattern_details/widgets/rich_markdown_body.dart:97]: Convert this to expression body (arrow function).

File: lib/widgets/image_view.dart [hunk 2]
```diff
@@ -0,0 +1,145 @@
+import 'dart:async';
+
+import 'package:flutter/material.dart';
```
Comment (mkobuolys) [inline lib/widgets/image_view.dart:3]: Import sorting is wrong.

File: lib/widgets/image_view.dart [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+  @override
+  State<ImageView> createState() => _ImageViewState();
+}
+
+class _ImageViewState extends State<ImageView>
+    with SingleTickerProviderStateMixin {
+  late TransformationController _controller;
+  late Animation<Matrix4>? _animationReset;
+  late AnimationController _controllerReset;
+  Offset _tapPosition = Offset.zero;
```
Comment (mkobuolys) [inline lib/widgets/image_view.dart:22]: Add an empty line before this.

File: lib/widgets/image_view.dart [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+    super.dispose();
+  }
+
+  void _onAnimateReset() {
+    _controller.value = _animationReset!.value;
+    if (!_controllerReset.isAnimating) {
+      _animationReset?.removeListener(_onAnimateReset);
+      _animationReset = null;
+      _controllerReset.reset();
+    }
```
Comment (mkobuolys) [inline lib/widgets/image_view.dart:49]:
```suggestion
    if (_controllerReset.isAnimating) return;
    
      _animationReset?.removeListener(_onAnimateReset);
      _animationReset = null;
      _controllerReset.reset();
```

File: lib/widgets/image_view.dart [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+  }
+
+  void _onDoubleTap() {
+    if (isInitial) {
+      _animateScaleIn();
+    } else {
+      _animateScaleOut();
+    }
+  }
```
Comment (mkobuolys) [inline lib/widgets/image_view.dart:58]:
```suggestion
  void _onDoubleTap() => isInitial ? _animateScaleIn() : _animateScaleOut();
```

File: lib/widgets/image_view.dart [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+
+  // little hack as there is no way to get onTapDown triggered
+  // when onDoubleTap is registered
+  // track issue: (https://github.com/flutter/flutter/issues/10048)
+  void _onTapDown(TapDownDetails details) {
+    if (_tapPosition != Offset.zero &&
```
Comment (mkobuolys) [inline lib/widgets/image_view.dart:64]: If case body is empty - invert the condition to avoid this.

File: lib/widgets/image_view.dart [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+        (details.globalPosition.dx - _tapPosition.dx < 20 &&
+            details.globalPosition.dy - _tapPosition.dy < 20)) {
+    } else {
+      _tapPosition = details.globalPosition;
+    }
+    Timer(const Duration(milliseconds: 500), () {
```
Comment (mkobuolys) [inline lib/widgets/image_view.dart:70]: Add space before this.

File: lib/widgets/image_view.dart [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+      _tapPosition = Offset.zero;
+    });
+  }
+
+  void _animateScaleIn() {
+    final offset = Offset(
+      -MediaQuery.of(context).size.width / 4,
+      -MediaQuery.of(context).size.height / 4,
+    );
+    _animateResetInitialize(
```
Comment (mkobuolys) [inline lib/widgets/image_view.dart:80]: Add space before this.

File: lib/widgets/image_view.dart [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+    Timer(const Duration(milliseconds: 500), () {
+      _tapPosition = Offset.zero;
+    });
+  }
+
+  void _animateScaleIn() {
+    final offset = Offset(
+      -MediaQuery.of(context).size.width / 4,
+      -MediaQuery.of(context).size.height / 4,
+    );
```
Comment (mkobuolys) [inline lib/widgets/image_view.dart:79]:
```suggestion
    final size = MediaQuery.of(context).size;
    final offset = Offset(-size.width / 4, -size.height / 4);
```

File: lib/widgets/image_view.dart [hunk 10]
```diff
@@ -0,0 +1,145 @@
...
+    _controllerReset.reset();
+    _animationReset = Matrix4Tween(
+      begin: _controller.value,
+      end: end,
+    ).animate(_controllerReset);
+    _animationReset?.addListener(_onAnimateReset);
+    _controllerReset.forward();
+  }
+
+  void _animateResetStop() {
```
Comment (mkobuolys) [inline lib/widgets/image_view.dart:101]: The code seems like a duplicate of `_onAnimateReset` - extract similar logic to a dedicated function and reuse it.

File: lib/widgets/image_view.dart [hunk 11]
```diff
@@ -0,0 +1,145 @@
...
+    _animationReset = null;
+    _controllerReset.reset();
+  }
+
+  void _onInteractionStart(ScaleStartDetails details) {
+    // If the user tries to cause a transformation while the reset animation is
+    // running, cancel the reset animation.
+    if (_controllerReset.status == AnimationStatus.forward) {
+      _animateResetStop();
+    }
```
Comment (mkobuolys) [inline lib/widgets/image_view.dart:113]:
```suggestion
    if (_controllerReset.status != AnimationStatus.forward) return;
    
    _animateResetStop();    
```

File: lib/widgets/image_view.dart [hunk 12]
```diff
@@ -0,0 +1,145 @@
...
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    return Scaffold(
+      appBar: AppBar(
+        elevation: 0,
+        backgroundColor: lightBackgroundColor,
+        leading: const PlatformBackButton(color: Colors.black),
```
Comment (mkobuolys) [inline lib/widgets/image_view.dart:122]: I don't think that the PlatformBackButton is the right one to use there. I would prefer to have a "X" button on the top right.

**Review by mkobuolys — CHANGES_REQUESTED** [review]
My initial feeling is that the `ImageView` is a bit overengineered and not everything that's implemented there is needed. Anyway, let's refactor the code a bit for better readability, I will dive a bit deeper with the other iteration.

**Comment by Juliotati** [comment]
> My initial feeling is that the `ImageView` is a bit overengineered and not everything that's implemented there is needed. Anyway, let's refactor the code a bit for better readability, I will dive a bit deeper with the other iteration.

Considering the code age it's most likely that that's the case

**Pull request merged**